components/brave_perf_predictor/browser/p3a_bandwidth_savings_permanent_state.cc

```diff
@@ -0,0 +1,117 @@
+
+#include <numeric>
+#include <utility>
+
+#include "brave/components/brave_perf_predictor/common/pref_names.h"
+#include "components/prefs/pref_registry_simple.h"
+#include "components/prefs/pref_service.h"
+#include "components/prefs/scoped_user_pref_update.h"
+
+namespace brave_perf_predictor {
+
+namespace {
+
+constexpr size_t kNumOfSavedDailyUptimes = 7;
+
+}  // namespace
+
+P3ABandwidthSavingsPermanentState::P3ABandwidthSavingsPermanentState(
+    PrefService* user_prefs)
+    : user_prefs_(user_prefs) {
+  if (user_prefs)
+    LoadSavingsDaily();
+}
+
+P3ABandwidthSavingsPermanentState::~P3ABandwidthSavingsPermanentState() =
+    default;
+
+void P3ABandwidthSavingsPermanentState::AddSavings(uint64_t delta) {
+  base::Time now_midnight = base::Time::Now().LocalMidnight();
+  base::Time last_saved_midnight;
+
+  if (!daily_savings_.empty())
+    last_saved_midnight = daily_savings_.front().day;
+
+  if (now_midnight - last_saved_midnight > base::TimeDelta()) {
+    // Day changed.
+    daily_savings_.emplace_front(DailySaving{now_midnight, delta});
+    if (daily_savings_.size() > kNumOfSavedDailyUptimes)
+      daily_savings_.pop_back();
+  } else {
+    daily_savings_.front().saving += delta;
+  }
+
+  SaveSavingsDaily();
+}
+
+base::Optional<uint64_t>
+P3ABandwidthSavingsPermanentState::GetFullPeriodSavingsBytes() {
+  if (daily_savings_.size() == kNumOfSavedDailyUptimes) {
+    // divide by 1024*1024 = 2^20 to convert bytes -> MB
+    return GetSavingsTotal() >> 20;
+  } else {
+    return base::nullopt;
+  }
+}
+
+uint64_t P3ABandwidthSavingsPermanentState::GetSavingsTotal() const {
+  // We record only saving for last N days.
+  const base::Time n_days_ago =
+      base::Time::Now() - base::TimeDelta::FromDays(kNumOfSavedDailyUptimes);
+  return std::accumulate(daily_savings_.begin(), daily_savings_.end(), 0UL,
+                         [n_days_ago](const uint64_t acc, const auto& u2) {
+                           uint64_t add = 0;
+                           // Check only last continious days.
+                           if (u2.day > n_days_ago) {
+                             add = u2.saving;
+                           }
+                           return acc + add;
+                         });
+}
+
+void P3ABandwidthSavingsPermanentState::LoadSavingsDaily() {
+  DCHECK(daily_savings_.empty());
+  if (!user_prefs_)
+    return;
+  const base::ListValue* list =
+      user_prefs_->GetList(prefs::kBandwidthSavedDailyBytes);
+  if (!list)
+    return;
+
+  for (auto it = list->begin(); it != list->end(); ++it) {
+    const base::Value* day = it->FindKey("day");
+    const base::Value* saving = it->FindKey("saving");
+    if (!day || !saving || !day->is_double() || !saving->is_double())
+      continue;
+    if (daily_savings_.size() == kNumOfSavedDailyUptimes)
+      break;
+    daily_savings_.emplace_back(
+        DailySaving{base::Time::FromDoubleT(day->GetDouble()),
```

Comment:
`daily_savings_.emplace_back(base::Time::FromDoubleT(day->GetDouble()), static_cast<uint64_t>(saving->GetDouble()));`
and for this you need a constructor. 
Otherwise just use `push_back` here